### DISCUSSION TITLE
DNG to FITS (FTS = Flexible Image Transport System) format conversion

### DISCUSSION THREAD

**Discussion by TobyOnGithub** [ImageMagick/ImageMagick#7049] [General]
I am using the astrophotography app "DeepSkyCameraBeta" which generates DNG files. How to use imagemagick to best convert to FITS for astronomical image processing without loss of information?

**Comment by snibgo** [comment]
```
magick in.dng out.fits
```
That should convert the image. It may not copy all metadata that doesn't affect the image (such as camera details).

**Reply by TobyOnGithub** [reply]
hmmm I figured out that instead of "magick" command I should use "convert". But I get this error:

laptop:~/Desktop$ convert Bias_0001_20231227_131504123000000.dng out.fits
convert-im6.q16: delegate failed `'ufraw-batch' --silent --create-id=also --out-type=png --out-depth=16 --output='%u.png' '%i'' @ error/delegate.c/InvokeDelegate/1966.
convert-im6.q16: unable to open image `/tmp/magick-_ibxNuYxiAlCRVqOPziUuPbPNoYoPUWv.ppm': No such file or directory @ error/blob.c/OpenBlob/2924.
convert-im6.q16: no images defined `out.fits' @ error/convert.c/ConvertImageCommand/3229.

**Comment by snibgo** [comment]
You don't have `magick`, so your version of IM is probably old. I suggest you upgrade to v7, which probably uses libraw to read DNG files.

Alternatively, you might install `ufraw-batch`.

**Reply by TobyOnGithub** [reply]
One step in the right direction, but not quite. I installed magick v7 (configure, make, sudo make install). First without libraw installation. Encountered errors related to ufraw-batch. Then I also installed libraw (I assume libraw is better than ufraw-batch) and rebuild magick and installed it. Somewhere along the configuration I could see a libraw path. No errors when I test magick v7. Then I convert a DNG using your suggestion. It now runs without an error, but the resulting FITS file is not a cube (the DNG comes from a smartphone app specialized on astrophotography) and the image displayed (in the SAOIMAGE/DS9 FITS image viewer) shows 65535 for all pixels!! I appreciate your help and hope you (or anyone else) have a suggestion to solve this ... :-) ... maybe I do should install ufraw-batch? But it is no longer supported I think.....

edit: For clarification, the image I am trying to convert is a BIAS image. So it should not be 65535 for all pixels.

**Comment by snibgo** [comment]
I don't do astrophotography, and don't know what a "cube" file is, or "BIAS".

What is the text output of `magick -version`? I hope it includes HDRI, so IM can process values outside the normal 0-100% range.

The problem may be reading the DNG, or writing the FITS.

What is the text output from:
```
magick in.dng -verbose info:
```
Near the top of the text output it will list the minimum and maximum values in all the channels. Are they all `65535 (1.0)` or are they sensible values?

Try:
```
magick in.dng -define quantum:format=floating-point -depth 32 -auto-level out.tiff
```
This should write a 32 bit/channel floating-point TIFF file. `-auto-level` will adjust values to the mimimum is zero and the maximum is 100%. The result should be a visible image. Some old viewers misread 32-bit FP, so if it looks wrong, try removing `-define quantum:format=floating-point -depth 32`.

If that does work okay, then perhaps the problem is writing the FITS. That format can also use `-define quantum:format=floating-point -depth 32`, see https://www.imagemagick.org/script/formats.php, so try that.

Can you zip or input DNG and post it here?

**Reply by TobyOnGithub** [reply]
thank you so much for your feedbacks!!

answer to "what is the text output from magick -version":

```
laptop:~/Desktop$ magick -version
Version: ImageMagick 7.1.1-27 (Beta) Q16-HDRI x86_64 83eefaf2a:20240107 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): fontconfig freetype jbig jng jpeg ltdl lzma openexr pangocairo png raw tiff x xml zlib
Compiler: gcc (11.4)
```

answer to "magick in.dng -verbose info:"

```
laptop:~/Desktop$ magick Bias_0001_20231227_131504123000000.dng -verbose info:
Image:
  Filename: Bias_0001_20231227_131504123000000.dng
  Format: DNG (Digital Negative)
  Class: DirectClass
  Geometry: 4032x3024+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: Bilevel
  Base type: Undefined
  Endianness: Undefined
  Depth: 16/1-bit
  Channels: 3.0
  Channel depth:
    Red: 1-bit
    Green: 1-bit
    Blue: 1-bit
  Channel statistics:
    Pixels: 12192768
    Red:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      median: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: 0
      skewness: 0
      entropy: 0
    Green:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      median: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: 0
      skewness: 0
      entropy: 0
    Blue:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      median: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: 0
      skewness: 0
      entropy: 0
  Image statistics:
    Overall:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      median: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: 0
      skewness: 0
      entropy: 0
  Colors: 1
  Histogram:
      12192768: (65535,65535,65535) #FFFFFFFFFFFF white
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33,0.03)
    green primary: (0.3,0.6,0.1)
    blue primary: (0.15,0.06,0.79)
    white point: (0.3127,0.329,0.3583)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: black
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 4032x3024+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Undefined
  Orientation: RightTop
  Properties:
    date:create: 2024-01-18T14:45:26+00:00
    date:modify: 2023-12-27T12:15:04+00:00
    date:timestamp: 2024-01-20T13:34:21+00:00
    dng:camera.model.name: SM-G981B
    dng:create.date: 2023-12-27T12:15:03+00:00
    dng:exposure.time: 1/1e+02
    dng:f.number: 2
    dng:focal.length: 5 mm
    dng:focal.length.in.35mm.format: 0 mm
    dng:gps.altitude: 0 m
    dng:gps.latitude: 0 deg 0' 0" N
    dng:gps.longitude: 0 deg 0' 0" W
    dng:iso.setting: 0
    dng:lens: 0-0mm f/0-0
    dng:lens.f.stops: 0.00
    dng:lens.type: 
    dng:make: Samsung
    dng:max.aperture.at.max.focal: 0
    dng:max.aperture.at.min.focal: 0
    dng:max.aperture.value: 0
    dng:max.focal.length: 0 mm
    dng:min.focal.length: 0.0 mm
    dng:software: samsung/x1sxeea/x1s:13/TP1A.220624.014/G981BXXSIHWKD:user/relea
    dng:wb.rb.levels: 1.97303 1.54917 1 0
    signature: 6251541af00b8fa2d40552c7de985abdc1c6bfe0f1de452d7cf2522f96b2c085
  Tainted: False
  Filesize: 23.2807MiB
  Number pixels: 12.1928M
  Pixel cache type: Memory
  Pixels per second: 31.9248GP
  Time-to-live: 0:0:0:0  2024-01-20T13:34:21Z
  User time: 0.000u
  Elapsed time: 0:01.000
  Version: ImageMagick 7.1.1-27 (Beta) Q16-HDRI x86_64 83eefaf2a:20240107 https://imagemagick.org
```

answer to ""

```
laptop:~/Desktop$ magick Bias_0001_20231227_131504123000000.dng -define quantum:format=floating-point -depth 32 -auto-level out.tiff
```

This creates a out.tiff file without any errors. When opening in GIMP, I see a uniform white image. Perhaps all pixels have maximum values? The BIAS image is just a 0-second exposure image recording the background noise of the CCD for later subtraction. It should not have all pixels maxed out.

answer to "removing -define quantum:format=floating-point -depth 32":

this again produces a out.tiff file. When loaded with GIMP again the image is just white.

```
[Bias_0001_20231227_131504123000000.dng.zip](https://github.com/ImageMagick/ImageMagick/files/13998121/Bias_0001_20231227_131504123000000.dng.zip)
```

I have uploaded a zipped file of the DNG file.

**Comment by snibgo** [comment]
Your `magick -version` says you have HDRI and `raw` which means it uses libraw to read the DNG. That's fine.

For your zipped file:

- `magick` (using libraw) says it is entirely white.

- `ufraw` (I don't know if it uses libraw) says it is entirely white.

- `dcraw_emu` (using libraw) makes a TIFF file that `magick` says it is entirely white.

- `darktable` (which doesn't use libraw) shows an image that looks entirely white.

I conclude that the image in your DNG file is entirely white. You say, "It should not have all pixels maxed out." Okay, but I'm afraid it is entirely white.

DNG files often contain JPG thumbnails, made by the camera. This is a useful sanity check. However, your DNG has no thumbnails.

**Comment by snibgo** [comment]
After some more work: your raw file does contain slightly varied data, after all. It contains 12192768 pixels, each of a single channel. The values are 16-bit. On a scale of 0 to 65535, 12178286 pixels (99.88% of the pixels) have a value of 1023. The rest have values between 729 and 1022. This is a very small variation. The mean value is 1022.94.

The trick is that these numbers are only visible in the raw data, while it is still mosaiced (ie Bayer). The de-mosaicing process smooths the differences to almost nothing. The programs I used all de-mosaic the data, but we shouldn't do that.

That makes sense. The data in the file represents the value at each sensel (sensor pixel) of the "no-light" exposure. This data shouldn't be de-mosaiced. Instead, it should be subtracted (or something) from each ordinary photo of stars, *before* demosaicing those photos.

We can get the raw data in four files like this:
```
4channels.exe -B Bias_0001_20231227_131504123000000.dng
```
`4channels.exe` is part of the libraw package. For each file, we can examine the pixel values with `magick`.

ImageMagick always de-mosaics raw image files. So it is the wrong software to get the data you want.

**Comment by TobyOnGithub** [comment]
Thank you so much for all this invaluable information. Can I ask you to look at this project:

https://github.com/olegkutkov/Raw2Fits

I managed to build raw2fits with libraw. It runs nicely. However, I am not sure what settings to use to transfer original measurements to the resulting FITS file(s). The documentation is good, but not in-depth. However, "apply interpolation" should be unchecked according to your last remark that we should not de-mosaic raw images. Can I ask you to run raw2fits on the lat DNG file that I shared in this thread?

**Comment by snibgo** [comment]
For your dark frame, I suggest: don't interpolate, don't autobright, and don't autoscale.

See also https://github.com/LibRaw/LibRaw/blob/master/samples/dcraw_emu.cpp . This program will read a raw file that contains a photograph, and the `-K <file>` option is to "subtract dark frame". I think your file `Bias_0001_20231227_131504123000000.dng` is the dark frame. From the dark frame, we can create a grayscale PGM file:
```
unprocessed_raw.exe C:\web\im\Bias_0001_20231227_131504123000000.dng
```
Then `magick` can read the PGM file, and give use information about it, or save to another format:
```
magick Bias_0001_20231227_131504123000000.dng.pgm -verbose info:

  Channel statistics:
    Pixels: 12192768
    Gray:
      min: 729  (0.0111238)
      max: 1023 (0.01561)
      mean: 1022.94 (0.0156091)
      median: 1023 (0.01561)
      standard deviation: 2.18629 (3.33606e-05)
      kurtosis: 3373.31
      skewness: -53.0139
      entropy: 0.00271087
```
The PGM file can be used in a `dcraw_emu` program, option `-K`.

The `git clone` doesn't work for me. Perhaps this is because I have a very slow internet connection. In about a week, I will have a faster connection.
```
c:\cygwin64\home\Alan\imagemagick-7.1.1-4B\ImageMagick-Windows-7\ImageMagick>git clone https://github.com/olegkutkov/Raw2Fits.git
Cloning into 'Raw2Fits'...
remote: Enumerating objects: 1188, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (8/8), done.
error: 5181 bytes of body are still expectediB | 329.00 KiB/s
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output

c:\cygwin64\home\Alan\imagemagick-7.1.1-4B\ImageMagick-Windows-7\ImageMagick>git clone https://github.com/olegkutkov/Raw2Fits.git
Cloning into 'Raw2Fits'...
remote: Enumerating objects: 1188, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (8/8), done.
error: 797 bytes of body are still expected MiB | 360.00 KiB/s
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output

c:\cygwin64\home\Alan\imagemagick-7.1.1-4B\ImageMagick-Windows-7\ImageMagick>git clone https://github.com/olegkutkov/Raw2Fits.git
Cloning into 'Raw2Fits'...
remote: Enumerating objects: 1188, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (8/8), done.
error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
error: 792 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```
